### PR TITLE
Fix issue where TimeDistributed doesn't pass uses_learning_phase

### PR DESCRIPTION
**Pull request by EntilZha**
It seems that the `TimeDistributed` layer has a similar issue as the `Bidirectional` layer when passed a layer that requires the learning phase as described in:

* https://github.com/fchollet/keras/issues/5940
* https://github.com/fchollet/keras/issues/5975
* https://github.com/fchollet/keras/pull/5985

This simple example reproduces the problem and is included in the unit tests in the PR. This was tested on a clean copy of the master branch on the commit prior to the PR.

```python
from keras.layers import core, wrappers
from keras.models import Sequential
import numpy as np

model = Sequential()
model.add(wrappers.TimeDistributed(core.Dropout(.5), input_shape=(3, 2)))
model.compile(optimizer='rmsprop', loss='mse')
model.predict(np.random.random((10, 3, 2)))
```

The error output is longish so I put it in this gist: https://gist.github.com/EntilZha/06a8bbd2279d01a61a1c7a5e61d882b4

My proposed fix mimics this section of code from `Bidirectional`, but applies it to `TimeDistributed`: https://github.com/fchollet/keras/blob/master/keras/layers/wrappers.py#L257-L285

I tested my changes locally using `py.test tests` before my changes and after my changes and all tests pass. `py.test --pep8 -m pep8` also passes without errors. The machines I am using use tensorflow `1.0.0-rc0` and `1.0.0`.

Thanks!

### CODE REVIEW

**Comment by EntilZha** [comment]
I just realized I was a bit careless and although those tests pass, this doesn't seem to return the correct result, namely that the mask weight is applied to everything, rather than each time step individually:

```python
model = Sequential()
model.add(TimeDistributed(Dropout(.5), input_shape=(3, 2)))
K.set_learning_phase(1)
model.compile(optimizer='rmsprop', loss='mse')
x = np.array([
    [[1, 1], [2, 2], [-1, -1]]
])
model.predict(x)
```

Output
```
array([[[ 2.,  2.],
        [ 4.,  0.],
        [-2., -2.]]], dtype=float32)
```

File: keras/layers/wrappers.py [hunk 1]
```diff
@@ -184,6 +189,10 @@ def step(x, _):
            self.layer.activity_regularizer is not None):
             regularization_loss = self.layer.activity_regularizer(y)
             self.add_loss(regularization_loss, inputs)
+
+        if 'training' in kwargs:
+            y._uses_learning_phase = True
```
Comment (fchollet) [inline keras/layers/wrappers.py]: This is incorrect in the general case: just because `call` takes a `training` argument doesn't mean the learning phase is being used. Instead, you should look at whether `_uses_learning_phase` is set on any if the output tensors of the child layer.

File: tests/keras/layers/wrappers_test.py [hunk 2]
```diff
@@ -86,6 +86,12 @@ def test_TimeDistributed():
     outer_model.compile(optimizer='rmsprop', loss='mse')
     outer_model.fit(np.random.random((10, 3, 2)), np.random.random((10, 3, 3)), epochs=1, batch_size=10)
 
+    # test layers that need learning_phase to be set
+    model = Sequential()
+    model.add(wrappers.TimeDistributed(core.Dropout(.5), input_shape=(3, 2)))
+    model.compile(optimizer='rmsprop', loss='mse')
+    model.predict(np.random.random((10, 3, 2)))
```
Comment (fchollet) [inline tests/keras/layers/wrappers_test.py]: This test doesn't seem to test for correctness...

File: keras/layers/wrappers.py (hunk 1 above)
Comment (EntilZha) [inline keras/layers/wrappers.py]: Seems like a relatively easy fix, I'll change that

File: tests/keras/layers/wrappers_test.py (hunk 2 above)
Comment (EntilZha) [inline tests/keras/layers/wrappers_test.py]: I am thinking a good test would be something like below, but account for the cases where dropout eliminates the entire input so the sum would be zero

```python
x = np.array([
    [[1, 1], [2, 2], [-1, -1]]
])
y = model.predict(x)
np.testing.assert_allclose(x.sum(axis=2), y.sum(axis=2), atol=1e-05)
```

To get that to pass though I would need to sort out the mask issue. It's unclear to me if it's a model issue (my fault), or `TimeDistributed` doesn't yet handle masking correctly in all cases (https://github.com/fchollet/keras/issues/4786 and https://github.com/fchollet/keras/issues/5212)

**Comment by EntilZha** [comment]
Thanks for the comments on PR. In going through what I thought was the issue with dropout not being computed correctly I realized that it was indeed being computed correctly, and my understanding was incorrect. Moreover, I think that `TimeDistributed(Dropout))` and just `Dropout` on the prior layer would have the same effect.

So, in this specific case I don't think it makes sense to need `TimeDistributed`, but the bug I ran into is related to the layer needing the learning phase, not necessarily it being dropout. I'll go ahead and make the requested changes in both how to test for learning phase being needed as well as improving the test.

**Comment by fchollet** [comment]
Do you have any update on this PR?

**Comment by EntilZha** [comment]
For what I was trying to accomplish, turns out it's not necessary since I had a misunderstanding with how dropout worked, namely that doing `TimeDistributed(Dropout)` was no better than just `Dropout` (at least I think).

I haven't had time recently to work on this, but could come back to it after this month if there is still value in doing it for other layers that need the use training phase parameter passed through.

**Comment by fchollet** [comment]
@EntilZha is the PR abandoned?
@AziziShekoofeh Would you like to take over the PR?

**Comment by EntilZha** [comment]
@fchollet I would gladly let @AziziShekoofeh take over. Its not blocking any of my current work so I am not sure when I would get around to finishing it.

**Comment by AziziShekoofeh** [comment]
Sure, I am still trying to find a way to solve it.

**Comment by fchollet** [comment]
Fixed it and merged it.

**Comment by abnera** [comment]
@fchollet Thanks. I thought this would fix TimeDistributed error with BatchNormalization or Dropout layers since they seem to have a similar issue, but sadly we cannot still use TimeDistributed with BatchNormalization/Dropout (https://github.com/fchollet/keras/issues/5934).

Any trick for using TimeDistributed or Reshape to get a sequence of images `(num_samples, seq_length, h, w, c)`  pass each image in seq_length via the cnn_model and then stack the cnn outputs `(num_samples, seq_length, features)`. Only works if no BatchNormalization and/or Dropout in the CNN model, but most deep model need this layers to avoid overfitting during training.
